ulog/context.go

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package ulog
+
+import (
+	"context"
+	"fmt"
+
+	opentracing "github.com/opentracing/opentracing-go"
+	jaeger "github.com/uber/jaeger-client-go"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+// Logger returns a *zap.Logger that's pre-seeded with tracing information from
+// the provided context.
+func Logger(ctx context.Context) *zap.Logger {
+	if ctx == nil {
+		return zap.L()
+	}
+	return zap.L().With(Trace(ctx))
+}
+
+// Sugar retrieves a loosely-typed zap.SugaredLogger from the provided context.
+func Sugar(ctx context.Context) *zap.SugaredLogger {
+	return Logger(ctx).Sugar()
+}
+
+// Trace creates a field that extracts tracing information from a context
+// (including trace, span, and parent IDs) and includes it under the "trace"
+// key.
+//
+// Because the opentracing APIs don't expose this information, the returned
+// zap.Field is a no-op for contexts that don't contain a span or contain a
+// non-Jaeger span.
+func Trace(ctx context.Context) zapcore.Field {
+	if ctx == nil {
+		return zap.Skip()
+	}
+	return zap.Object("trace", trace{ctx})
+}
+
+type trace struct {
+	ctx context.Context
+}
+
+func (t trace) MarshalLogObject(enc zapcore.ObjectEncoder) error {
+	span := opentracing.SpanFromContext(t.ctx)
+	if span == nil {
+		return nil
+	}
+	j, ok := span.Context().(jaeger.SpanContext)
+	if !ok {
+		return nil
```

Comment:
Can you elaborate a bit? If there's no Jaeger span available, there's nothing to do; it's not an error, it's an expected case.